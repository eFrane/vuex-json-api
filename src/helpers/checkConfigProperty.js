import { hasOwn } from '../shared/utils'

/**
 *
 * @param {object} config
 * @param {String} property
 * @param {Boolean} isRequiredProp
 */
export function checkConfigProperty (config, property, isRequiredProp = true) {
  if (config !== null && hasOwn(config, property)) {
    return true
  }

  if ((config === null ||
    hasOwn(config, property) === false) &&
    !isRequiredProp) {
    return false
  }

  throw new Error('Missing configuration property: ' + property)
}
