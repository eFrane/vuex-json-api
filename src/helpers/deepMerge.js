/**
 * Performs a deep merge of `source` into `target`.
 * Mutates `target` only but not its objects and arrays.
 *
 * @author inspired by [jhildenbiddle](https://stackoverflow.com/a/48218209).
 *
 * @param {Object} target
 * @param {Object} source
 *
 * @return {Object} merged Orbject
 */

export function deepMerge (target, source) {
  const isObject = (obj) => obj && typeof obj === 'object'

  if (!isObject(target) || !isObject(source)) {
    return source
  }

  Object.keys(source).forEach(key => {
    const targetValue = target[key]
    const sourceValue = source[key]

    if (Array.isArray(targetValue) && Array.isArray(sourceValue)) {
      target[key] = targetValue.concat(sourceValue)
    } else if (isObject(targetValue) && isObject(sourceValue)) {
      target[key] = deepMerge(Object.assign({}, targetValue), sourceValue)
    } else {
      target[key] = sourceValue
    }
  })

  return target
}
