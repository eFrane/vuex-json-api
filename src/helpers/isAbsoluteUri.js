/**
 *
 * @param {String} uri
 */
export function isAbsoluteUri (uri) {
  if (uri.indexOf('http://') === 0) {
    return true
  }

  if (uri.indexOf('https://') === 0) {
    return true
  }

  if (uri.match(/\/\/.+/)) {
    return true
  }

  return false
}
