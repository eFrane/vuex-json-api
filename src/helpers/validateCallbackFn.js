export function validateCallbackFn (fn) {
  if (fn.constructor !== Function) {
    return false
  }

  return true
}
