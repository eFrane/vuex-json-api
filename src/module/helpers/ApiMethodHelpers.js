import { hasOwn } from '../../shared/utils'

function checkPropertyOnObject (obj, property) {
  return typeof obj === 'object' && hasOwn(obj, property)
}

/**
 * Check whether the api supports collections
 *
 * @param {object} apiMethods supported api methods (list, get, update, delete)
 */
function isCollection (apiMethods) {
  return checkPropertyOnObject(apiMethods, 'list')
}

function allowsModification (apiMethods) {
  return checkPropertyOnObject(apiMethods, 'update')
}

function allowsReplacement (apiMethods) {
  return checkPropertyOnObject(apiMethods, 'replace')
}

function allowsCreation (apiMethods) {
  return checkPropertyOnObject(apiMethods, 'create')
}

function allowsDeletion (apiMethods) {
  return checkPropertyOnObject(apiMethods, 'delete')
}

function hasRelated (apiMethods) {
  return checkPropertyOnObject(apiMethods, 'related')
}

export {
  allowsCreation,
  allowsDeletion,
  allowsModification,
  allowsReplacement,
  hasRelated,
  isCollection
}
