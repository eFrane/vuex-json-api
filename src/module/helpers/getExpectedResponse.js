function getExpectedResponse (currentResource) {
  if (!currentResource.id && !currentResource.type) {
    throw new Error('Resources must have an id-property and a type-property.')
  }
  return {
    ...Object.assign(
      currentResource, {
        id: currentResource.id,
        type: currentResource.type
      }
    )
  }
}

export { getExpectedResponse }
