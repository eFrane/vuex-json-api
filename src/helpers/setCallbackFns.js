import { validateCallbackFn } from './validateCallbackFn'

/**
 *
 * @param {Array} callbacks
 */
export function setCallbackFns (callbacks) {
  if (typeof callbacks === 'undefined' ||
    callbacks.constructor !== Array ||
    callbacks.reduce((carry, cb) => validateCallbackFn(cb) && carry, true) === false) {
    throw new Error('You must pass an array of valid callback functions to this method')
  }

  return callbacks
}
