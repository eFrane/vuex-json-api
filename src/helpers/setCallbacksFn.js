import {validateCallbackFn} from './validateCallbackFn'

/**
 *
 * @param {Array} callbacks
 */
export function setCallbacksFn (callbacks) {
  if (typeof callbacks === 'undefined' || callbacks.constructor !== Array ||
    callbacks.reduce((carry, cb) => validateCallbackFn(cb) && carry, true) === false) {
    throw new Error('You must pass an array of callbacks to this method')
  }
  return callbacks
}
