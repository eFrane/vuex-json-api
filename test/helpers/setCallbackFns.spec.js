import { setCallbackFns } from '@/helpers/setCallbackFns'

test('returns valid arrays', () => {
  const cbArr1 = [
    () => {}
  ]

  const cbArr2 = [
    (foo) => { return foo },
    (bar) => { return bar }
  ]

  expect(setCallbackFns(cbArr1)).toStrictEqual(cbArr1)
  expect(setCallbackFns(cbArr2)).toStrictEqual(cbArr2)
})

test('throws on invalid arrays', () => {
  const cbArr3 = [
    (foo) => { return foo },
    42,
    (bar) => { return bar }
  ]

  function testInvalid () {
    setCallbackFns(cbArr3)
  }

  expect(testInvalid).toThrowErrorMatchingSnapshot()
})
