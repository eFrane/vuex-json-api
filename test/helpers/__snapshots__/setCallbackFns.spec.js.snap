// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setCallbackFns throws on invalid arrays 1`] = `"You must pass an array of valid callback functions to this method"`;

exports[`throws on invalid arrays 1`] = `"You must pass an array of valid callback functions to this method"`;
